In [12]:
import numpy as np
import pandas as pd
from sklearn import datasets, linear_model
import matplotlib.pyplot as plt
import os
%matplotlib inline

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.decomposition import PCA
from sklearn.decomposition import NMF

In [13]:
data = pd.read_csv('User Data Processed.csv')
data.head()

User ID  User Difficulty Choice  \
0  ID790145788                     1.0   
1  ID646234447                     0.0   
2   ID86918707                     5.0   
3  ID404262345                     2.0   
4  ID267634255                     4.0   

                     User Read Books (2017)  \
0                                      6254   
1                    5854, 5676, 7265, 6051   
2                                      5815   
3                          6969, 7002, 5761   
4  7908, 7585, 7655, 5013, 5352, 7753, 5793   

                           book contain words (2017)  User Read number (2017)  \
0  ['3368', '6030', '5392', '4091', '1495', '2018...                        1   
1  ['4602', '8051', '2572', '1107', '489', '5096'...                        4   
2  ['7064', '4813', '489', '5611', '71', '3122', ...                        1   
3  ['5011', '4546', '6419', '2018', '4680', '7302...                        3   
4  ['5035', '1838', '3524', '7466', '7911', '6356...                        7   

                               User Read shop (2017)  \
0                                        PythonBooks   
1  EduBooks, PythonBooks, Level 5 Daniel Shop, Le...   
2                                  Daniel's Bookshop   
3                    EduBooks, PythonBooks, EduBooks   
4  Level 5 Daniel Shop, Level 5 Daniel Shop, Leve...   

              User Read price (2017)      User Read genre (2017)  \
0                                6.0                          17   
1                 1.0, 3.0, 8.0, 7.0                 0, 0, 0, 21   
2                                5.0                           4   
3                      5.0, 2.0, 0.0                  19, 19, 17   
4  6.0, 9.0, 1.0, 1.0, 3.0, 5.0, 8.0  14, 14, 14, 14, 14, 29, 29   

  User Read sales (2017) User Read Words (2017)          ...           \
0                      6                      9          ...            
1             6, 4, 5, 1             2, 8, 6, 5          ...            
2                      6                      0          ...            
3                0, 5, 0                7, 7, 2          ...            
4    5, 4, 3, 3, 4, 1, 2    7, 1, 8, 6, 5, 0, 0          ...            

                           User Read Books (2018)  \
0                                            7180   
1                    5729, 5279, 6242, 5519, 5251   
2                                      6372, 6697   
3                                5023, 6412, 6124   
4  7915, 7884, 7404, 6022, 6589, 6543, 5450, 6848   

                           book contain words (2018) User Read number (2018)  \
0  ['5611', '4201', '3996', '2934', '140', '2036'...                       1   
1  ['5418', '398', '2572', '6716', '2202', '4454'...                       5   
2  ['4377', '1389', '633', '3524', '3606', '1381'...                       2   
3  ['4984', '5004', '8133', '5418', '8090', '8080...                       3   
4  ['5187', '4749', '8076', '6127', '5863', '5418...                       8   

                               User Read shop (2018)  \
0                                         DHC-Online   
1  PythonBooks, Level 5 Daniel Shop, Level 5 Dani...   
2                               DHC-Online, EduBooks   
3                  EduBooks, PythonBooks, DHC-Online   
4  EduBooks, EduBooks, EduBooks, DHC-Online, DHC-...   

                   User Read price (2018)          User Read genre (2018)  \
0                                     1.0                              18   
1                 5.0, 5.0, 8.0, 8.0, 1.0              21, 21, 16, 16, 16   
2                                3.0, 2.0                            4, 4   
3                           4.0, 2.0, 8.0                        17, 4, 4   
4  1.0, 8.0, 9.0, 1.0, 6.0, 8.0, 6.0, 8.0  29, 29, 29, 25, 25, 25, 25, 25   

   User Read sales (2018)  User Read Words (2018) User Read Difficulty (2018)  \
0                       3                       1                           1   
1           6, 3, 0, 1, 3           1

### 2017年图书的文字处理

In [31]:
book_words_2017 = data['book contain words (2017)']

In [103]:
book_words_2017.shape

(10000,)

- **要再写个循环，用“，”隔开，再join成list。但是那样的话矩阵会很大，没有必要这么做吧？**
- **所以如果是文字，用赋予topic/情感值的方式，会更有解释意义，维度也会大幅减小**

In [68]:
len(book_words_2017[0])

35488

In [89]:
len('3368')

4

In [69]:
len(['3368'])

1

In [105]:
book_words_2017[:5]

0    ['3368', '6030', '5392', '4091', '1495', '2018...
1    ['4602', '8051', '2572', '1107', '489', '5096'...
2    ['7064', '4813', '489', '5611', '71', '3122', ...
3    ['5011', '4546', '6419', '2018', '4680', '7302...
4    ['5035', '1838', '3524', '7466', '7911', '6356...
Name: book contain words (2017), dtype: object

In [104]:
book_words_2017.values[:5]

array(["['3368', '6030', '5392', '4091', '1495', '2018', '778', '1472', '6100', '6584', '6419', '6183', '3226', '6785', '5418', '1388', '946', '1651', '5587', '4377', '140', '2934', '2934', '5960', '4056', '1001', '5061', '140', '8006', '7512', '7064', '6274', '3917', '1380', '6419', '940', '6477', '7137', '5703', '3122', '7639', '2303', '3849', '6048', '7348', '5401', '1651', '5613', '140', '6419', '6419', '6370', '1287', '4586', '4246', '3401', '1001', '2934', '6506', '3555', '7953', '5187', '6506', '7064', '5343', '5004', '5835', '4471', '7502', '4377', '3524', '5309', '6907', '4377', '6239', '140', '7208', '5004', '2113', '2532', '4234', '6419', '6419', '5075', '6819', '1275', '5004', '7302', '47', '2875', '5187', '71', '4415', '5035', '3524', '5835', '140', '4377', '5849', '2658', '1287', '3046', '1001', '2934', '5418', '524', '4546', '3076', '2173', '7880', '5418', '6239', '1688', '6506', '5418', '2934', '3069', '5418', '32', '71', '140', '6346', '7602', '3728', '4889', '4377', '

**CounterVectorizer：记录每行中每个单词在所有单词中出现的次数**。会得到很稀疏的矩阵，10000*所有单词的个数（比35488维要大），而大多数位置都是0.

** It takes a loooong time...**

https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html

- lowercase : boolean, True by default. 这里全是数字，所以要给改成False。
- tokenizer : callable or None (default). Override the string tokenization step while preserving the preprocessing and n-grams generation steps. Only applies if analyzer == 'word'. 不是word是数字，所以用匿名函数lamda映射成本身

In [106]:
counter = CountVectorizer(lowercase=False, tokenizer= lambda x: x)
#counter.fit(book_words_2017)
counter.fit(book_words_2017.values)

In [107]:
print(counter.get_feature_names())

[' ', "'", ',', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '[', ']']


In [108]:
len(counter.get_feature_names())

15

In [109]:
# todense是稠密表示，是有很多0的表示
# toarray returns an ndarray; todense returns a matrix. 
# book_words_2017_count = counter.transform(book_words_2017).todense()
book_words_2017_count = counter.transform(book_words_2017)

In [35]:
book_words_2017_count.shape

(10000, 15)

In [39]:
# 稀疏矩阵的表现形式
print(book_words_2017_count[0])

  (0, 0)	4521
  (0, 1)	9044
  (0, 2)	4521
  (0, 3)	1235
  (0, 4)	2205
  (0, 5)	1778
  (0, 6)	2168
  (0, 7)	2224
  (0, 8)	1875
  (0, 9)	1513
  (0, 10)	1740
  (0, 11)	1411
  (0, 12)	1251
  (0, 13)	1
  (0, 14)	1


In [40]:
book_words_2017_count_dense = book_words_2017_count.todense()

In [41]:
book_words_2017_count_dense[0]

matrix([[4521, 9044, 4521, 1235, 2205, 1778, 2168, 2224, 1875, 1513,
         1740, 1411, 1251,    1,    1]])

#### 只剩下15个feature，不太对啊。。

CountVector会把str中的每个词分出来，而如果str中如项目中，文字是用数字来代替的话，CountVector就会把一个一个的数字切出来，而不是把一串数字看作一个整体，就导致了只有数字加符号的15维这种问题。所以，显然，数字是不能用CountVector的。


- Transform a **count matrix** to a normalized tf or tf-idf representation
- smooth_idf : boolean, default=True。Smooth idf weights by adding one to document frequencies, as if an extra document was seen containing every term in the collection exactly once. **Prevents zero divisions**. 因为这里只有一个文章，所以设为False。

In [44]:
transformer = TfidfTransformer(smooth_idf=False)
book_words_2017_tf_idf = transformer.fit_transform(book_words_2017_count_dense)
book_words_2017_tf_idf = book_words_2017_tf_idf.toarray()

In [50]:
book_words_2017_tf_idf.shape

(10000, 15)

In [49]:
book_words_2017_tf_idf[0]

array([3.64010051e-01, 7.28181133e-01, 3.64010051e-01, 9.94364993e-02,
       1.77536422e-01, 1.43156353e-01, 1.74557353e-01, 1.79066214e-01,
       1.50966345e-01, 1.21819776e-01, 1.40096768e-01, 1.13607207e-01,
       1.00724745e-01, 8.05153840e-05, 8.05153840e-05])

In [51]:
# PCA
# pca = PCA(n_components=1000)
# pca.fit(book_words_2017_counter)
# book_words_2017_pca = pca.transform(book_words_2017_counter)
# book_words_2017_pca.shape

#n_components=1000 must be between 0 and n_features=15 with svd_solver='full'

- **Non-Negative Matrix Factorization (NMF)**
- Find two non-negative matrices (W, H) whose product approximates the non- negative matrix X. This factorization can be used for example for **dimensionality reduction, source separation or topic extraction**.
- ‘random’: non-negative random matrices, scaled with: **sqrt(X.mean() / n_components)**
- https://blog.csdn.net/ACdreamers/article/details/44663421

### 正常来说，应该

<img src='https://images.cnitblog.com/blog/571227/201412/291500574341861.png'>

**m=10000,n=15,r=100.这是降的什么维啊。。。**

In [53]:
# NMF
nmf = NMF(n_components=100, init='random', random_state=0)
W = nmf.fit_transform(book_words_2017_count_dense)
H = nmf.components_

In [54]:
# 基矩阵
W.shape

(10000, 100)

In [57]:
W[0]

array([2.13292403e+01, 4.00717865e+00, 0.00000000e+00, 8.82469555e+01,
       4.32381999e-01, 0.00000000e+00, 6.56493471e-01, 0.00000000e+00,
       0.00000000e+00, 2.01274243e+00, 2.51062183e+01, 3.65291675e+01,
       0.00000000e+00, 1.65672491e+00, 1.21236320e-01, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 6.56296403e+01, 4.29408718e-02,
       2.64808773e-01, 0.00000000e+00, 1.58920055e+00, 0.00000000e+00,
       2.85085440e+00, 2.86129372e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       8.41873096e-01, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 5.43119124e+01, 4.01979693e-01, 1.06056049e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 6.24611457e+00,
       7.28973419e-02, 1.19081563e-01, 7.45057803e-02, 0.00000000e+00,
       0.00000000e+00, 1.59340684e+01, 1.34006334e+01, 0.00000000e+00,
       3.37651726e-01, 0.00000000e+00, 9.84501707e-02, 9.72774526e-01,
      

In [58]:
# 系数矩阵
H.shape

(100, 15)

### 2018年图书的文字处理

In [59]:
book_words_2018 = data['book contain words (2018)']
counter = CountVectorizer(lowercase=False, tokenizer= lambda x: x)
counter.fit(book_words_2018)
book_words_2018 = counter.transform(book_words_2018).todense()

In [70]:
# TF-IDF DEPRECATED
transformer = TfidfTransformer(smooth_idf=False)
book_words_2018_tf_idf = transformer.fit_transform(book_words_2018)
book_words_2018_tf_idf = book_words_2018_tf_idf.toarray()

In [71]:
# PCA for 2018
# book_words_2018 = book_words_2018.todense()
# 'matrix' object has no attribute 'todense'

# pca = PCA(n_components=1000)
# pca.fit(book_words_2018)
# book_words_2018_pca = pca.transform(book_words_2018)
# book_words_2018_pca.shape

In [72]:
# NMF For 2018
nmf = NMF(n_components=100, init='random', random_state=0)
W_2018 = nmf.fit_transform(book_words_2018)
H_2018 = nmf.components_

In [73]:
W_2018.shape

(10000, 100)

### One-hot

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from scipy.sparse import csr_matrix
def one_hotted(colomn_name):
    result = data[colomn_name].str.split(', ')
    counter = CountVectorizer(lowercase=False, tokenizer= lambda x: x)
    counter.fit(result)
    result = counter.transform(result).todense()
#     result[result>1] = 1
    return result

In [ ]:
# start one hottlize
user_choice = one_hotted('User Difficulty Choice')

user_slaes_2017 = one_hotted('User Read sales (2017)')
user_words_2017 = one_hotted('User Read Words (2017)')
user_difficulty_2017 = one_hotted('User Read Difficulty (2017)')
user_genre_2017 = one_hotted('User Read genre (2017)')
user_price_2017 = one_hotted('User Read price (2017)')
user_shop_2017 = one_hotted('User Read shop (2017)')
user_num_2017 = one_hotted('User Read number (2017)')
# book_words_2017 = one_hotted('book contain words (2017)')

user_slaes_2018 = one_hotted('User Read sales (2018)')
user_words_2018 = one_hotted('User Read Words (2018)')
user_difficulty_2018 = one_hotted('User Read Difficulty (2018)')
user_genre_2018 = one_hotted('User Read genre (2018)')
user_price_2018 = one_hotted('User Read price (2018)')
user_shop_2018 = one_hotted('User Read shop (2018)')
user_num_2018 = one_hotted('User Read number (2018)')
# book_words_2018 = one_hotted('book contain words (2018)')

In [ ]:
# user_slaes_2017.shape
# user_words_2017.shape
# user_difficulty_2017.shape
# user_genre_2017.shape
# user_price_2017.shape
# user_shop_2017.shape
user_num_2017.shape
# user_choice.shape

### 小知识：

- **Sparse Matrix**

In [18]:
from numpy import array
from scipy.sparse import csr_matrix

In [20]:
# create dense matrix
A = array([[1, 0, 0, 1, 0, 0], [0, 0, 2, 0, 0, 1], [0, 0, 0, 2, 0, 0]])
# convert to sparse matrix (CSR method)
S = csr_matrix(A)
S
# 把原来的18个element变成5个

<3x6 sparse matrix of type '<class 'numpy.int64'>'
	with 5 stored elements in Compressed Sparse Row format>

In [21]:
print(S)

  (0, 0)	1
  (0, 3)	1
  (1, 2)	2
  (1, 5)	1
  (2, 3)	2


In [22]:
# 对稀疏化之后的矩阵进行还原
B = S.todense()
B

matrix([[1, 0, 0, 1, 0, 0],
        [0, 0, 2, 0, 0, 1],
        [0, 0, 0, 2, 0, 0]], dtype=int64)

In [24]:
len(B)

3

In [26]:
B.shape

(3, 6)

In [ ]:
# 对正常矩阵进行todense（）
B_dense = B.todense()

- **CountVectorizer**

In [60]:
from sklearn.feature_extraction.text import CountVectorizer
corpus = [
'This is the first document.',
'This document is the second document.',
'And this is the third one.',
'Is this the first document?',
]
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(corpus)
print(vectorizer.get_feature_names())

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


**CountVector会把str中的每个词分出来，而如果str中如项目中，文字是用数字来代替的话，CountVector就会把一个一个的数字切出来，而不是把一串数字看作一个整体，就导致了只有数字加符号的15维这种问题。所以，显然，数字是不能用CountVector的。**

In [63]:
X

<4x9 sparse matrix of type '<class 'numpy.int64'>'
	with 21 stored elements in Compressed Sparse Row format>

In [78]:
# 第0行，9列元素分别有什么的稀疏表示
print(X[0])

  (0, 1)	1
  (0, 2)	1
  (0, 6)	1
  (0, 3)	1
  (0, 8)	1


In [64]:
X.todense()

matrix([[0, 1, 1, 1, 0, 0, 1, 0, 1],
        [0, 2, 0, 1, 0, 1, 1, 0, 1],
        [1, 0, 0, 1, 1, 0, 1, 1, 1],
        [0, 1, 1, 1, 0, 0, 1, 0, 1]], dtype=int64)

**针对数字的CountVector**

In [80]:
numbers = [
    ['3368', '6030', '5392', '4091', '1495', '2018'],
    ['4602', '8051', '2572', '1107', '489', '5096'],
    ['7064', '4813', '489', '5611', '71', '3122'],
    ['5011', '4546', '6419', '2018', '4680', '7302'],
    ['5035', '1838', '3524', '7466', '7911', '6356']
]

In [84]:
vectorizer = CountVectorizer(lowercase=False,tokenizer= lambda x: x)
X = vectorizer.fit_transform(numbers)
X

<5x28 sparse matrix of type '<class 'numpy.int64'>'
	with 30 stored elements in Compressed Sparse Row format>

In [88]:
print(vectorizer.get_feature_names())

['1107', '1495', '1838', '2018', '2572', '3122', '3368', '3524', '4091', '4546', '4602', '4680', '4813', '489', '5011', '5035', '5096', '5392', '5611', '6030', '6356', '6419', '7064', '71', '7302', '7466', '7911', '8051']


In [86]:
print(X[0])

  (0, 3)	1
  (0, 1)	1
  (0, 8)	1
  (0, 17)	1
  (0, 19)	1
  (0, 6)	1


In [87]:
X.todense()

matrix([[0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
         0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
         0, 1, 1, 0, 0, 0, 0],
        [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
         1, 0, 0, 1, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
         0, 0, 0, 0, 1, 1, 0]], dtype=int64)